eth/handler.go

```diff
@@ -288,22 +292,68 @@ func newHandler(config *handlerConfig) (*handler, error) {
 	return h, nil
 }
 
+// protoTracker tracks the number of active protocol handlers.
+func (h *handler) protoTracker() {
+	defer h.wg.Done()
+	var (
+		active int
+		done   = make(chan struct{})
+	)
+	// Start goroutine to track active protocol handlers.
+	go func() {
+		for {
+			select {
+			case <-h.handlerStartCh:
+				active--
+			case <-h.handlerDoneCh:
+				active++
+			case <-done:
+				return
+			}
+		}
+	}()
+	<-h.quitSync
+
+	// After quit is initiated, poll until all handlers are closed.
+	for range time.Tick(time.Millisecond * 50) {
+		if active == 0 {
+			break
+		}
+	}
```

Comment:
This is one way to do it, sure. Another way is to use a `sync.Cond`, this snippet would just `Wait()` on the cond, and the loop above would Signal or Broadcast on it.

Reply:
I tried a bit with `sync.Cond` but it seemed like a lot of overhead and `sync.WaitGroup` seems more straightforward. I have never actually successfully pulled off a `sync.Cond` though so it's possibly simpler still than what I came up with.